### ISSUE TITLE
Known issues with multiplayer

### ISSUE THREAD

**Issue by Discontinuum** [Dugy/Legend_of_the_Invincibles#300]
1. https://github.com/Dugy/Legend_of_the_Invincibles/issues/298
2. Other after-crafting options (smashing, drop on the ground) also cause OoS
3. "Pick up from the ground" causes OoS: other players don't see the item picked, for them it's still on the ground
4. Enemy units cause item events (picking a gem, gold, an item, though gems go to the players' storage and items' fate decided by an mp host [he got an item window])

**Comment by edwardspec** [comment]
_(disclaimer: I don't have capability to test the game in multiplayer, so I can't fix this myself, but the information below should be useful for whoever tries to fix it)_

By its design, the Inventory dialog itself is resistant to OoS errors, because it does every item-related operation via
```lua
inventory_dialog.mpsafety:queue({ ... })
```
... rather than invoking them directly.

The catch is, **any item-affecting changes** that happen **while the player has the Inventory dialog open** should also be performed via `mpsafety:queue()`. Otherwise in multiplayer these changes will be performed in another order for other players.

Both Crafting and "Pick up from the ground" are calling an event "item_pick", and that event is not using `mpsafety.queue()`. This is what causes OoS.

Simplest solution is probably to explose the variable `inventory_dialog.mpsafety` (make it not `local`) to be usable from outside the Inventory dialog's code (and use it in event handler of `item_pick`).

**Comment by Discontinuum** [comment]
I have found the reason for the 4th issue: enemies picking items in MP. The trick is that an item dropped from a defeated enemy is added with the "loti.item.on_the_ground.add" facility from lua/items.lua. And when it adds an event handler for picking the item, it uses filter_side "controller=human". But according to https://wiki.wesnoth.org/StandardSideFilter , this filter cannot be used in multiplayer games. That's why in MP enemies pick items while in SP everything's fine

**Comment by Discontinuum** [comment]
Found
5. Gems transmuting doesn't work, other players don't see changes

**Comment by Discontinuum** [comment]
Well, the situation from I see is the following. Any information others can get from the moment a player opened the Inventory menu is from the mpsafety queue. Any items, like item_pick, fired in the dialogue are fired only for the local player. Crafting is absolutely unsynched, numbers of gems are unsynched. Crafting in the game is working with a crutch "craft" > "store" > "equip", but it works only because the game doesn't check if the item to equip exists. So, when you craft and store, for others this item doesn't exist in their storages, but when you eqiup from your local storage, this event comes into the mpsafety queue, other folks repeat the operation of "equipping", for them the equipped item is created out of thin air.

Basically we need two things: make crafting operations be done via `mpsafety:queue()`, and make `item_pick` event run with it either. But for the latter one I'm not sure, the mpsafety queue is used only for the inventory dialogue, so inserting into `item_pick` used elsewhere (not only in the inventory dialogue) seems troublesome. Write a distinctive `item_pick_queued` with mpsafety for use in the inventory dialogue? That seems a code duplication, not cool

**Comment by Dugy** [comment]
I think your analysis is correct. The items menu was written by edwardspec who had the MP safety in mind, but I wrote the crafting menu before the items menu was written, not realising it was causing OoS. Crafting therefore should use the the `mp_safety:queue()`, but at the moment it doesn't because I didn't remember to write the mp safety code. So yes, it should be part of the multiplayer safety queue.

Regarding the `item_pick` duplication problem, can it be done in a way where the queued version calls the basic `item_pick`?

**Comment by Discontinuum** [comment]
`item_pick` does several things: it shows the menu and it contains the code implementing the options, taking, decomposing into a gem (quite long stuff tbh). For a queued implementation we don't want to have that implementation code be run right in place, we want to move implementation into `mpsafety` and in place we will only enlist our intentions.

A possible solution draft which comes to my mind is to make lua implementations of taking from the ground/dropping/destroying/leaving, which would use `mpsafety`. Then `item_pick`, which will be used only for picking items from the ground, will show the menu and run actions inside some new `run_mpsafe_operations` tag (the tag will run `synchronize` method of `mpsafety`). And for the inventory dialogue its `item_pick2` will show the menu and run operations to add their results into the used mpsafety queue (which would be flushed when the inventory dialogue closed). This way the duplication will be reduced only to creating menus. I wonder though if I'm overcomplicating anything right now.

**Comment by Dugy** [comment]
You are not overcomplicating anything, I think it's the right approach.

**Comment by Discontinuum** [comment]
With #338 merged the only bug remaining from the list here is 6. disabling attacks for retal not synchronized

**Comment by Discontinuum** [comment]
#342 fixes this for good
